src/effects/How to make an effect stateless.txt

```diff
@@ -0,0 +1,189 @@
+    
+            struct Instance;
+
+            std::shared_ptr<EffectInstance> MakeInstance(EffectSettings& settings) const;
+
+        Identify members that represent state - comment them out 
+        
+        remove methods: ProcessInitialize, ProcessFinalize, ProcessBlock
+
+    In cpp file,
+    
+        add:
+    
+        struct EffectEcho::Instance
+           : public PerTrackEffect::Instance
+           , public EffectInstanceWithBlockSize
+           , public EffectInstanceWithSampleRate
+
+        {
+           Instance(const PerTrackEffect& effect)
+              : PerTrackEffect::Instance{ effect }
+           {}
+
+           bool ProcessInitialize(EffectSettings& settings,
+              sampleCount totalLen, ChannelNames chanMap) override;
+
+           size_t ProcessBlock(EffectSettings& settings,
+              const float* const* inBlock, float* const* outBlock, size_t blockLen)  override;
+
+           bool ProcessFinalize(void) override;
+
+           << MEMBERS WHICH REPRESENT STATE which you commented out earlier >>
+        };
+
+
+        std::shared_ptr<EffectInstance>
+        EffectEcho::MakeInstance(EffectSettings&) const
+        {
+           return std::make_shared<Instance>(*this);
+        }
+        
+        
+    Make the 3 Effect::Process*** methods belong to the instance by: 
+    
+    - prepending ::Instance where needed
+
+    - adding this at the beginning (may not be needed in all of them)
+    
+        // temporary - in the final step this will be replaced by
+        // auto& echoSettings = GetSettings(settings);
+        //
+        auto& echoSettings = static_cast<const EffectEcho&>(mProcessor).mSettings;
+
+    
+
+
+4) "Real Statelessness"
+
+    In header file:
+    
+    - make the effect inherit  EffectWithSettings<EffectEchoSettings, PerTrackEffect>
+    
+    - remove the FetchParameters method
+    
+    - remove mSettings member
+
+
+    In cpp file:
+    
+    - go to processInitialize, ProcessBlock, ProcessFinalize; remove the downcast and enable the commented out GetSettings
+    
+    - go to Validator::ValidateUI & Validator::UpdateUI, remove the commenting out on the commented lines
+    
+    - add const to arg EffectEchoSettings& in the Validator constructor
+    
+    - turn Validator::mSettings from a reference to a non-reference
+    
+    - modify PopulateOrExchange to make it belong to the Validator
+      -- remove parameter EffectSettings from it, it is not needed anymore
+      -- it should now return void
+      
+    - reimplement Effect::PopulateOrExchange:
+    
+        std::unique_ptr<EffectUIValidator>
+        EffectEcho::PopulateOrExchange(ShuttleGui& S, EffectSettingsAccess& access)
+        {
+           auto& settings = access.Get();
+           auto& myEffSettings = GetSettings(settings);
+           auto result = std::make_unique<Validator>(*this, access, myEffSettings);
+           result->PopulateOrExchange(S);
+           return result;
+        }
+
```

Comment:
There may be a fifth step:  see the commit I recently pushed to the branch for ladspa effects.

`MakeSettings` and `CopySettingsContents` may require explicit definitions instead of those generated by the template EffectWithSettings.

That does not apply to built-in effects where the settings structure simply contains several scalar fields and no variable sized containers.

For Ladspa, you can see that MakeSettings() instead allocates a vector of values, whose size is not known at compilation time but depends on other information that the effect is queried for.  So default construction of the settings, as in the generated MakeSettings(), will not do.

Then CopySettingsContents must also be overridden, not for correctness but for efficiency.  The override is supposed to avoid memory allocations when the main thread calls it, so it should not use the copy constructors of the embedded std::vector but instead rewrite an existing vector without changing its capacity.  The assumption will be made that the destination settings containers are already correctly sized, because MakeSettings did that, or else full copies done in the main thread (where allocations are allowed) already did that.

Other third party effect families like AudioUnits may have Settings structures containing maps from strings to values, instead of vectors, but the problems will be similar.